File: homework/two.py
import collections
from collections.abc import Iterator
from datetime import datetime
from functools import cache
from multiprocessing import Pool, cpu_count
from multiprocessing.pool import Pool as IPool
from pathlib import Path
from typing import Literal, TypeAlias

from libs.slide_min import Index, SlideMin, Value

Minimizer: TypeAlias = int
RefId: TypeAlias = str
ReadId: TypeAlias = str
Strand = Literal[0, 1]
Twist = Literal[0, 1]

CURRENT_FOLDER_DIR = Path(__file__).resolve().parent

NUCLEOTIDE_TO_INTEGER_MAPPING: dict[
    Literal["A", "C", "G", "T"], Literal[0, 1, 2, 3]
] = {"A": 0, "C": 1, "G": 2, "T": 3}


class Minimap:
    read_seq_file_path: Path
    ref_seq_file_path: Path
    mapping_result_output_file_path: Path
    window_size: int
    k_mer_size: int
    k_mer_hash_mod: int
    processes: int

    def __init__(
        self,
        read_seq_file_path: Path,
        ref_seq_file_path: Path,
        window_size: int,
        k_mer_size: int,
        mapping_result_output_file_path: Path | None = None,
        processes: int | None = None,
    ) -> None:
        self.read_seq_file_path = read_seq_file_path
        self.ref_seq_file_path = ref_seq_file_path
        self.mapping_result_output_file_path = (
            CURRENT_FOLDER_DIR
            / f'mapping_result_PacBio_{datetime.now().isoformat(timespec="seconds")}.tsv'
            if mapping_result_output_file_path is None
            else mapping_result_output_file_path
        )
        self.k_mer_size = k_mer_size
        self.k_mer_hash_mod = (
            1 << ((k_mer_size - 1) * 2)
        ) - 1  # = 4 ** (k_mer_size - 1) - 1, 下位bit全て1
        self.window_size = window_size
        self.processes = (
            cpu_count() if processes is None else min(processes, cpu_count())
        )
        print(f"This program is runnig with {self.processes} processes.")

    def parse_ref_seq_file(self) -> dict[RefId, str]:
        ref_seq_dict: dict[RefId, str] = {}
        with self.ref_seq_file_path.open("r") as f:
            tmp_ref_id: RefId | None = None
            seq = ""

            while True:
                line = f.readline().rstrip()
                if len(line) == 0:
                    break
                if line.startswith(">"):
                    if seq != "":
                        ref_seq_dict[tmp_ref_id] = seq
                        seq = ""
                    tmp_ref_id = line.split()[0][1:]
                else:
                    seq += line
            ref_seq_dict[tmp_ref_id] = seq
        return ref_seq_dict

    def parse_read_seq_file(self) -> dict[ReadId, str]:
        read_seq_dict: dict[ReadId, str] = {}
        with self.read_seq_file_path.open("r") as f:
            while True:
                read_id = f.readline().rstrip()[1:]
                read_array = f.readline().rstrip()
                f.readline()
                f.readline()
                if len(read_id) == 0:
                    break
                read_seq_dict[read_id] = read_array
        return read_seq_dict

    def sketch_minimizers(self, seq: str) -> Iterator[tuple[Value, Index, Strand]]:
        window_size = self.window_size
        k_mer_size = self.k_mer_size
        minimizers_set: set[Index] = set()
        seq_length = len(seq)
        k_mer_init = seq[:k_mer_size]
        k_mer_hashes_strand_0 = SlideMin(self.seq2hash(k_mer_init), window_size)
        k_mer_hashes_strand_1 = SlideMin(
            self.seq2hash(k_mer_init, strand=1), window_size
        )
        for i in range(window_size - 1):
            last_k_mer_hash_strand_0_value = k_mer_hashes_strand_0.last_value
            last_k_mer_hash_strand_1_value = k_mer_hashes_strand_1.last_value
            (
                new_k_mer_hash_strand_0_value,
                new_k_mer_hash_strand_1_value,
            ) = self.calc_hash(
                last_k_mer_hash_strand_0_value,
                last_k_mer_hash_strand_1_value,
                seq[k_mer_size + i],
            )
            k_mer_hashes_strand_0.add(new_k_mer_hash_strand_0_value)
            k_mer_hashes_strand_1.add(new_k_mer_hash_strand_1_value)
        else:
            (
                k_mer_hash_strand_0_index,
                k_mer_hash_strand_0_value,
            ) = k_mer_hashes_strand_0.min
            (
                k_mer_hash_strand_1_index,
                k_mer_hash_strand_1_value,
            ) = k_mer_hashes_strand_1.min
            if k_mer_hash_strand_0_value < k_mer_hash_strand_1_value:
                minimizers_set.add(k_mer_hash_strand_0_index)
                yield k_mer_hash_strand_0_value, k_mer_hash_strand_0_index, 0
            if k_mer_hash_strand_0_value > k_mer_hash_strand_1_value:
                minimizers_set.add(k_mer_hash_strand_1_index)
                yield k_mer_hash_strand_1_value, k_mer_hash_strand_1_index, 1

        for j in range(seq_length - k_mer_size - window_size + 1):
            last_k_mer_hash_strand_0_value = k_mer_hashes_strand_0.last_value
            last_k_mer_hash_strand_1_value = k_mer_hashes_strand_1.last_value
            (
                new_k_mer_hash_strand_0_value,
                new_k_mer_hash_strand_1_value,
            ) = self.calc_hash(
                last_k_mer_hash_strand_0_value,
                last_k_mer_hash_strand_1_value,
                seq[k_mer_size + window_size + j - 1],
            )
            (
                k_mer_hash_strand_0_index,
                k_mer_hash_strand_0_value,
            ) = k_mer_hashes_strand_0.add(new_k_mer_hash_strand_0_value)

            (
                k_mer_hash_strand_1_index,
                k_mer_hash_strand_1_value,
            ) = k_mer_hashes_strand_1.add(new_k_mer_hash_strand_1_value)
            if (
                k_mer_hash_strand_0_value < k_mer_hash_strand_1_value
                and k_mer_hash_strand_0_index not in minimizers_set
            ):
                minimizers_set.add(k_mer_hash_strand_0_index)
                yield k_mer_hash_strand_0_value, k_mer_hash_strand_0_index, 0
            if (
                k_mer_hash_strand_0_value > k_mer_hash_strand_1_value
                and k_mer_hash_strand_1_index not in minimizers_set
            ):
                minimizers_set.add(k_mer_hash_strand_1_index)
                yield k_mer_hash_strand_1_value, k_mer_hash_strand_1_index, 1

    def generate_analyzed_read_seq_output(
        self,
        args: tuple[
            ReadId,
            str,
            collections.defaultdict[Minimizer, list[tuple[RefId, Index, Strand]]],
        ],
    ) -> str:
        read_id, read_seq, ref_minimizer_dict = args
        read_seq_length = len(read_seq)
        hits: collections.defaultdict[
            RefId, collections.Counter
        ] = collections.defaultdict(collections.Counter)
        read_minimizer_iterator = self.sketch_minimizers(read_seq)
        for (read_minimizer, read_pos, read_strand) in read_minimizer_iterator:
            for (ref_id, ref_pos, ref_strand) in ref_minimizer_dict[read_minimizer]:
                if read_strand == ref_strand:
                    hits[ref_id][ref_pos - read_pos + 1] += 1
                else:
                    hits[ref_id][
                        -(ref_pos - (read_seq_length - self.k_mer_size - read_pos) + 1)
                    ] += 1

        max_cnt = 0
        output = ""
        for ref_id, counter in hits.items():
            most_common_pos, most_common_cnt = counter.most_common(1)[0]
            if most_common_cnt > max_cnt:
                max_cnt = most_common_cnt
                output = self.format(
                    read_id, ref_id, most_common_pos, int(most_common_pos < 0)
                )
        return output

    def run(self) -> None:
        ref_seq_dict: dict[RefId, str] = self.parse_ref_seq_file()
        read_seq_dict: dict[ReadId, str] = self.parse_read_seq_file()

        ref_minimizer_dict: collections.defaultdict[
            Minimizer, list[tuple[RefId, Index, Strand]]
        ] = collections.defaultdict(list)
        for ref_id, ref_seq in ref_seq_dict.items():
            ref_minimizer_iterator = self.sketch_minimizers(ref_seq)
            for (minimizer, ref_pos, strand) in ref_minimizer_iterator:
                ref_minimizer_dict[minimizer].append((ref_id, ref_pos, strand))

        outputs: list[str] = []
        with Pool(processes=self.processes) as pool:
            pool: IPool
            outputs: list[str] = pool.map(
                self.generate_analyzed_read_seq_output,
                (
                    (read_id, read_seq, ref_minimizer_dict)
                    for read_id, read_seq in read_seq_dict.items()
                ),
            )

        with self.mapping_result_output_file_path.open("w") as f:
            f.writelines(outputs)

    def format(self, read_id: ReadId, ref_id: RefId, pos: int, strand_order: Twist):
        return f'{read_id}\t{ref_id}\t{abs(pos)}\t{"+-"[strand_order]}\n'

    def calc_hash(
        self,
        last_k_mer_hash_strand_0_value: Minimizer,
        last_k_mer_hash_strand_1_value: Minimizer,
        next_nucleotide: Literal["A", "C", "G", "T"],
    ) -> tuple[Minimizer, Minimizer]:
        k_mer_hash_mod = self.k_mer_hash_mod
        k_mer_hash_strand_0_value = (
            (last_k_mer_hash_strand_0_value & k_mer_hash_mod) << 2
        ) | self.seq2hash(next_nucleotide)
        k_mer_hash_strand_1_value = (last_k_mer_hash_strand_1_value >> 2) | (
            self.seq2hash(next_nucleotide, strand=1) << (2 * (self.k_mer_size - 1))
        )
        return k_mer_hash_strand_0_value, k_mer_hash_strand_1_value

    @staticmethod
    @cache
    def seq2hash(seq: str, *, strand: Strand = 0) -> int:
        res: int = 0
        if strand == 0:
            for v in seq:
                res = (res << 2) | NUCLEOTIDE_TO_INTEGER_MAPPING[v]
        else:
            for v in seq[::-1]:
                res = (res << 2) | (3 - NUCLEOTIDE_TO_INTEGER_MAPPING[v])
        return res


if __name__ == "__main__":
    IS_TEST = False
    if IS_TEST:
        minimap = Minimap(
            CURRENT_FOLDER_DIR / "test" / "PacBio.fastq",
            CURRENT_FOLDER_DIR / "test" / "ref.fasta",
            window_size=10,
            k_mer_size=20,
        )
        minimap.run()

    else:
        minimap = Minimap(
            CURRENT_FOLDER_DIR.parent / "SE11" / "PacBio_SE11.fastq",
            CURRENT_FOLDER_DIR.parent / "SE11" / "ref_SE11.fasta",
            window_size=10,
            k_mer_size=20,
        )
        minimap.run()

File: homework/two.cc
#include <algorithm>
#include <chrono>
#include <cstdlib>
#include <ctime>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <optional>
#include <queue>
#include <set>
#include <tuple>
#include <unordered_map>
#include <vector>
#include <time.h>

#include "libs/slide_min.h"

namespace fs = std::filesystem;
using libs::SlideMin;
using sc = std::chrono::system_clock;
using std::cout;
using std::endl;
using std::optional;
using std::string;
using std::tuple;
using std::unordered_map;
using std::vector;

using Minimizer = unsigned long long;
using RefId = string;
using ReadId = string;
enum class Strand : unsigned short
{
  Forward = 0,
  Reverse = 1
};

const unordered_map<char, short> NUCLEOTIDE_TO_INTEGER_MAPPING({{'A', 0}, {'C', 1}, {'G', 2}, {'T', 3}});
const std::nullopt_t null = std::nullopt;

class Minimap
{
private:
  const fs::path read_seq_file_path_;
  const fs::path ref_seq_file_path_;
  fs::path mapping_result_output_file_path_;
  const short window_size_;
  const short k_mer_size_;
  const unsigned long long k_mer_hash_mod_;

  unordered_map<RefId, string> ref_seq_dict;
  unordered_map<ReadId, string> read_seq_dict;
  unordered_map<Minimizer, vector<tuple<RefId, libs::Index, Strand>>> ref_minimizer_dict;
  // int processes_;

public:
  Minimap(
      const fs::path read_seq_file_path,
      const fs::path ref_seq_file_path,
      const short window_size,
      const short k_mer_size,
      optional<fs::path> mapping_result_output_file_path = null)
      : read_seq_file_path_(read_seq_file_path),
        ref_seq_file_path_(ref_seq_file_path),
        window_size_(window_size),
        k_mer_size_(k_mer_size),
        k_mer_hash_mod_((1ULL << ((k_mer_size - 1) * 2)) - 1)
  {
    time_t now;
    time(&now);
    char buf[sizeof "mapping_result_PacBio_2011-10-08T07:07:09.tsv"];
    strftime(buf, sizeof buf, "mapping_result_PacBio_%FT%T.tsv", gmtime(&now));
    mapping_result_output_file_path_ = (mapping_result_output_file_path.has_value()) ? mapping_result_output_file_path.value() : buf;
  }

  void parse_ref_seq_file()
  {
    std::ifstream ifs(ref_seq_file_path_);
    if (ifs.fail())
    {
      std::cerr << "Failed to open file." << std::endl;
      std::exit(1);
    }
    string tmp_ref_id;
    string seq = "";
    while (!ifs.eof())
    {
      string line;
      std::getline(ifs, line);
      if (line[0] == '>')
      {
        if (seq != "")
        {
          ref_seq_dict[tmp_ref_id] = seq;
          seq = "";
        }
        const short idx = line.find_first_of(' ');
        tmp_ref_id = line.substr(1, idx-1);
      }
      else
      {
        seq += line;
      }
    }
    ifs.close();
    ref_seq_dict[tmp_ref_id] = seq;
  }

  void parse_read_seq_file()
  {
    std::ifstream ifs(read_seq_file_path_);
    if (ifs.fail())
    {
      std::cerr << "Failed to open file." << std::endl;
      std::exit(1);
    }
    string tmp_ref_id;
    string seq = "";
    string _;
    while (true)
    {
      string read_id;
      std::getline(ifs, read_id);
      string read_array;
      std::getline(ifs, read_array);
      std::getline(ifs, _);
      std::getline(ifs, _);
      if (ifs.eof())
      {
        ifs.close();
        break;
      }
      read_seq_dict[read_id.substr(1)] = read_array;
    }
  }

  void sketch_minimizers(string &seq, std::queue<tuple<libs::Value, libs::Index, Strand>> &queue)
  {
    std::set<libs::Index> minimizers_set;
    const int seq_length = seq.size();
    const string k_mer_init = seq.substr(0, k_mer_size_);
    SlideMin k_mer_hashes_strand_0(seq2hash(k_mer_init), window_size_);
    SlideMin k_mer_hashes_strand_1(seq2hash(k_mer_init, Strand::Reverse), window_size_);

    for (short i = 0; i < window_size_ - 1; i++)
    {
      const Minimizer &last_k_mer_hash_strand_0_value = k_mer_hashes_strand_0.last_value();
      const Minimizer &last_k_mer_hash_strand_1_value = k_mer_hashes_strand_1.last_value();
      Minimizer new_k_mer_hash_strand_0_value;
      Minimizer new_k_mer_hash_strand_1_value;
      calc_hash(
          last_k_mer_hash_strand_0_value,
          last_k_mer_hash_strand_1_value,
          seq[k_mer_size_ + i],
          new_k_mer_hash_strand_0_value,
          new_k_mer_hash_strand_1_value);
      k_mer_hashes_strand_0.add(new_k_mer_hash_strand_0_value);
      k_mer_hashes_strand_1.add(new_k_mer_hash_strand_1_value);
    }
    const auto &[k_mer_hash_strand_0_index,
          k_mer_hash_strand_0_value] = k_mer_hashes_strand_0.min();
    const auto &[k_mer_hash_strand_1_index,
          k_mer_hash_strand_1_value] = k_mer_hashes_strand_1.min();
    if (k_mer_hash_strand_0_value < k_mer_hash_strand_1_value)
    {
      minimizers_set.insert(k_mer_hash_strand_0_index);
      queue.push(tuple(k_mer_hash_strand_0_value, k_mer_hash_strand_0_index, Strand::Forward));
    }
    if (k_mer_hash_strand_0_value > k_mer_hash_strand_1_value)
    {
      minimizers_set.insert(k_mer_hash_strand_1_index);
      queue.push(tuple(k_mer_hash_strand_1_value, k_mer_hash_strand_1_index, Strand::Reverse));
    }

    for (int j = 0; j < seq_length - k_mer_size_ - window_size_ + 1; j++)
    {
      const Minimizer &last_k_mer_hash_strand_0_value = k_mer_hashes_strand_0.last_value();
      const Minimizer &last_k_mer_hash_strand_1_value = k_mer_hashes_strand_1.last_value();
      Minimizer new_k_mer_hash_strand_0_value;
      Minimizer new_k_mer_hash_strand_1_value;
      calc_hash(
          last_k_mer_hash_strand_0_value,
          last_k_mer_hash_strand_1_value,
          seq[k_mer_size_ + window_size_ + j - 1],
          new_k_mer_hash_strand_0_value,
          new_k_mer_hash_strand_1_value);
      const auto &[k_mer_hash_strand_0_index,
            k_mer_hash_strand_0_value] = k_mer_hashes_strand_0.add(new_k_mer_hash_strand_0_value);
      const auto &[k_mer_hash_strand_1_index,
            k_mer_hash_strand_1_value] = k_mer_hashes_strand_1.add(new_k_mer_hash_strand_1_value);
      if (
          k_mer_hash_strand_0_value < k_mer_hash_strand_1_value && minimizers_set.count(k_mer_hash_strand_0_index) == 0)
      {
        minimizers_set.insert(k_mer_hash_strand_0_index);
        queue.push(tuple(k_mer_hash_strand_0_value, k_mer_hash_strand_0_index, Strand::Forward));
      }
      if (
          k_mer_hash_strand_0_value > k_mer_hash_strand_1_value && minimizers_set.count(k_mer_hash_strand_1_index) == 0)
      {
        minimizers_set.insert(k_mer_hash_strand_1_index);
        queue.push(tuple(k_mer_hash_strand_1_value, k_mer_hash_strand_1_index, Strand::Reverse));
      }
    }
  }

  string generate_analyzed_read_seq_output(ReadId read_id, string read_seq)
  {
    unsigned int read_seq_length = read_seq.size();
    unordered_map<RefId, unordered_map<libs::Index, unsigned short>> hits;
    std::queue<tuple<libs::Value, libs::Index, Strand>> read_minimizer_queue;
    sketch_minimizers(read_seq, read_minimizer_queue);
    while (!read_minimizer_queue.empty())
    {
      auto &[read_minimizer, read_pos, read_strand] = read_minimizer_queue.front();
      for (auto &[ref_id, ref_pos, ref_strand] : ref_minimizer_dict[read_minimizer])
      {
        if (read_strand == ref_strand)
        {
          hits[ref_id][ref_pos - read_pos + 1]++;
        }
        else
        {
          hits[ref_id][-(ref_pos - (read_seq_length - k_mer_size_ - read_pos) + 1)]++;
        }
      }
      read_minimizer_queue.pop();
    }

    short max_cnt;
    string output;
    for (auto &[ref_id, counter] : hits)
    {
      const auto &most_common_pair = std::max_element(
          counter.begin(), counter.end(), [](const auto &x, const auto &y)
          { return (x.second < y.second); });
      if (most_common_pair->second > max_cnt)
      {
        max_cnt = most_common_pair->second;
        output = read_id + '\t' + ref_id + '\t' + std::to_string(std::abs(most_common_pair->first)) + '\t' + "+-"[most_common_pair->first < 0];
      }
    }
    return output;
  }

  void calc_hash(
      Minimizer last_k_mer_hash_strand_0_value,
      Minimizer last_k_mer_hash_strand_1_value,
      char next_nucleotide,
      Minimizer &k_mer_hash_strand_0_value,
      Minimizer &k_mer_hash_strand_1_value)
  {
    k_mer_hash_strand_0_value = ((last_k_mer_hash_strand_0_value & k_mer_hash_mod_) << 2) | seq2hash(string(1, next_nucleotide), Strand::Forward);
    k_mer_hash_strand_1_value = (last_k_mer_hash_strand_1_value >> 2) | (seq2hash(string(1, next_nucleotide), Strand::Reverse) << (2 * (k_mer_size_ - 1)));
  }

  unsigned long long seq2hash(string seq, Strand strand = Strand::Forward)
  {
    unsigned long long res = 0;
    if (strand == Strand::Forward)
    {
      for (char &base : seq)
      {
        res = (res << 2) | NUCLEOTIDE_TO_INTEGER_MAPPING.at(base);
      }
    }
    else
    {
      for (short idx = seq.size(); idx-- > 0;)
      {
        char &base = seq[idx];
        res = (res << 2) | (3 - NUCLEOTIDE_TO_INTEGER_MAPPING.at(base));
      }
    }
    return res;
  }

  void run()
  {
    parse_ref_seq_file();
    parse_read_seq_file();

    std::queue<tuple<libs::Value, libs::Index, Strand>> ref_minimizer_queue;
    for (auto &ref_seq_pair : ref_seq_dict)
    {
      sketch_minimizers(ref_seq_pair.second, ref_minimizer_queue);
      while (!ref_minimizer_queue.empty())
      {
        auto &[minimizer, ref_pos, strand] = ref_minimizer_queue.front();
        ref_minimizer_dict[minimizer].push_back(tuple(ref_seq_pair.first, ref_pos, strand));
        ref_minimizer_queue.pop();
      }
    }

    std::ofstream ofs(mapping_result_output_file_path_);
    for (auto &[read_id, read_seq] : read_seq_dict)
    {
      string output = generate_analyzed_read_seq_output(read_id, read_seq);
      ofs << output << '\n';
    }
    ofs.close();
  }
};

int main()
{
  const bool IS_TEST = false;
  const fs::path CURRENT_FOLDER_DIR = fs::current_path();
  if (IS_TEST)
  {
    Minimap minimap(
        CURRENT_FOLDER_DIR / "homework" / "test" / "read.fastq",
        CURRENT_FOLDER_DIR / "homework" / "test" / "ref.fasta",
        10,
        15);
    minimap.run();
  }
  else
  {
    Minimap minimap(
        CURRENT_FOLDER_DIR / "SE11" / "PacBio_SE11.fastq",
        CURRENT_FOLDER_DIR / "SE11" / "ref_SE11.fasta",
        10,
        15);
    minimap.run();
  }

  return 0;
}
